src/main/java/smartthings/ratpack/protobuf/parsers/ProtobufParser.java

```diff
@@ -0,0 +1,60 @@
+package smartthings.ratpack.protobuf.parsers;
+
+import com.google.common.base.Strings;
+import com.google.common.reflect.TypeToken;
+import com.google.protobuf.Message;
+import com.google.protobuf.util.JsonFormat;
+import ratpack.handling.Context;
+import ratpack.http.TypedData;
+import ratpack.parse.NoOptParserSupport;
+import ratpack.util.Types;
+
+import java.lang.reflect.Method;
+
+import static smartthings.ratpack.protobuf.ContentType.*;
+
+/**
+ * Smartthings parsing of protocol buffers.
+ */
+public class ProtobufParser extends NoOptParserSupport {
+
+    private static final String CONTENT_TYPE_HEADER = "Content-Type";
+    private static final TypeToken<Message> PROTO3_TYPE_TOKEN = TypeToken.of(Message.class);
+
+    private JsonFormat.Parser jsonParser;
+
+    public ProtobufParser() {
+        jsonParser = JsonFormat.parser();
+    }
+
+    private static String getContentType(String header) {
+        if (Strings.isNullOrEmpty(header)) {
+            return JSON.getValue();
+        }
+        return header;
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    protected <T> T parse(Context ctx, TypedData requestBody, TypeToken<T> type) throws Exception {
+        if (!type.isSubtypeOf(PROTO3_TYPE_TOKEN)) {
+            return null;
+        }
+
+        String contentType = getContentType(ctx.getRequest().getHeaders().get(CONTENT_TYPE_HEADER));
+
+        if (PROTOBUF.getValue().equals(contentType)) {
+            Class<Message> clazz = (Class<Message>) type.getRawType();
+            Method method = clazz.getDeclaredMethod("parseFrom", byte[].class);
```

Comment:
I'm not sure how much this reflection stuff is actually impacting performance, but Charlie did prefer to have some caching on it a PR I made for a Dropwizard project.   You can see some of that here:
https://git.vandevlab.com/iot/execution/blob/master/execution-web/src/main/java/smartthings/execution/jersey/ProtocolBufferJsonMessageBodyProvider.java